src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs

```diff
@@ -921,45 +921,55 @@ public static bool IsValidParams(Symbol member)
         }
 
-        private static bool IsOverride(Symbol overridden, Symbol overrider)
+        /// <summary>
+        /// Does <paramref name="moreDerivedOverride"/> override <paramref name="member"/> or the
+        /// thing that it originally overrides, but in a more derived class?
+        /// </summary>
+        private static bool IsMoreDerivedOverride(Symbol member, Symbol moreDerivedOverride)
         {
-            if (TypeSymbol.Equals(overridden.ContainingType, overrider.ContainingType, TypeCompareKind.ConsiderEverything2) ||
-                !MemberSignatureComparer.SloppyOverrideComparer.Equals(overridden, overrider))
+            if (!moreDerivedOverride.IsOverride ||
+                !IsBaseClass(derivedType: moreDerivedOverride.ContainingType, baseType: member.ContainingType) ||
+                !MemberSignatureComparer.SloppyOverrideComparer.Equals(member, moreDerivedOverride))
             {
                 // Easy out.
                 return false;
             }
 
-            // Does overrider override overridden?
-            var current = overrider;
-            while (true)
-            {
-                if (!current.IsOverride)
-                {
-                    return false;
-                }
-                current = current.GetOverriddenMember();
+            // Rather than following the member.GetOverriddenMember() chain, we check to see if both
+            // methods ultimately override the same original method.  This addresses issues in binary compat
+            // scenarios where the override chain may skip some steps.
+            // See https://github.com/dotnet/roslyn/issues/45798 for an example.
+            return moreDerivedOverride.GetLeastOverriddenMember(accessingTypeOpt: null).OriginalDefinition ==
+                   member.GetLeastOverriddenMember(accessingTypeOpt: null).OriginalDefinition;
+        }
 
-                // We could be in error recovery.
-                if ((object)current == null)
-                {
-                    return false;
-                }
+        private static bool IsBaseClass(TypeSymbol derivedType, TypeSymbol baseType)
+        {
+            Debug.Assert((object)derivedType != null);
+            Debug.Assert((object)baseType != null);
 
-                if (current == overridden)
-                {
-                    return true;
-                }
+            // A base class has got to be a class. The derived type might be a struct, enum, or delegate.
+            if (!baseType.IsClassType())
+            {
+                return false;
+            }
 
-                // Don't search beyond the overridden member.
-                if (TypeSymbol.Equals(current.ContainingType, overridden.ContainingType, TypeCompareKind.ConsiderEverything2))
+            for (TypeSymbol b = derivedType.BaseTypeNoUseSiteDiagnostics; (object)b != null; b = b.BaseTypeNoUseSiteDiagnostics)
```

Comment:
>BaseTypeNoUseSiteDiagnostics [](start = 44, length = 28)

It is not clear why are we ignoring use-site diagnostics. It looks like problems in inheritance chain can affect the result. In cases like this we collect use-site diagnostics.

Reply:
We have always ignored use site diagnostics when determining if one method is a more specific override than another in overload resolution. This is not a change from existing compiler behavior.

---
In reply to: [461764045](https://github.com/dotnet/roslyn/pull/46367#discussion_r461764045) [](ancestors = 461764045)

Reply:
> We have always ignored use site diagnostics when determining if one method is a more specific override than another in overload resolution.

If that is the case, please point to the specific place in this component where use-site diagnostics is ignored when we traverse bases. 

> Use-site diagnostics would have been reported when gathering the candidates from the hierarchy of the receiver.

This component is not responsible to gather candidates and doesn't control how it is done. It can be called with any set of candidates as we more and more often build the set of candidates without relying on regular name binding from syntax.

---
In reply to: [461828701](https://github.com/dotnet/roslyn/pull/46367#discussion_r461828701) [](ancestors = 461828701,461764045)

Reply:
Note, that we have a Used Assembly References feature in the pipe line that depends on us always collecting use-site diagnostics every time there is a slightest chance that presence of the errors affects the analysis result in any way. We are quite tolerant to duplicate use-site errors reporting, but not as tolerant to missing recording an assembly "usage". I think it would be more robust long term if we simply collect the diagnostics here without trying to be too smart. 

---
In reply to: [461940796](https://github.com/dotnet/roslyn/pull/46367#discussion_r461940796) [](ancestors = 461940796,461828701,461764045)

Reply:
Those use site diagnostics have been reported when the candidates were collected.  I understand that you would prefer that overload resolution now be enhanced to be able to return use site diagnostics when computing the relationship between methods in the method group.  It did not do so before when traversing the inheritance hierarchy (when computing a method's overridden method) and I am not willing to do that in the context of this PR.  Please open a separate PR for that suggested work item.

---
In reply to: [461942977](https://github.com/dotnet/roslyn/pull/46367#discussion_r461942977) [](ancestors = 461942977,461940796,461828701,461764045)

Reply:
> If that is the case, please point to the specific place in this component where use-site diagnostics is ignored when we traverse bases. 

Original version of this code, line 941, (use-site diagnostics arising when traversing bases in order to compute the overridden member are discarded).

---
In reply to: [461940796](https://github.com/dotnet/roslyn/pull/46367#discussion_r461940796) [](ancestors = 461940796,461828701,461764045)